mesonbuild/mtest.py

```diff
@@ -1311,10 +1311,11 @@ def __init__(self, test: TestSerialisation, env: T.Dict[str, str], name: str,
             for c in self.cmd:
                 winecmd.append(c)
                 if os.path.basename(c).startswith('wine'):
-                    env['WINEPATH'] = get_wine_shortpath(
-                        winecmd,
-                        ['Z:' + p for p in self.test.extra_paths] + env.get('WINEPATH', '').split(';')
-                    )
+                    build_dir = os.getcwd()
+                    wine_paths = list(set(['Z:' + os.path.normpath(p) for p in self.test.extra_paths]
+                                          + env.get('WINEPATH', '').split(';')))
+                    wine_paths.sort(key=lambda x: not build_dir in x)
```

Comment:
I am not sure I understand how sorting these paths changes anything. Isn't the purpose simply to ensure that the `extra_paths` are placed first in `WINEPATH`? Doesn't the existing code already do that? I am confused.

Reply:
I can't even reproduce the bug you are talking about. I exported `WINEPATH="Z:\\usr\\x86_64-w64-mingw32\\bin;Z:\\usr\\x86_64-w64-mingw32\\sys-root\\mingw\\bin"` and built glib on Fedora against the system-wide cross-mingw64 toolchain and everything worked as expected.

Reply:
> I am not sure I understand how sorting these paths changes anything. 

Assume that `Z:\home\xxx\src\glib-mingw32-build` is your build directory, `Z:\usr\i686-w64-mingw32\\sys-root\mingw\bin` is the path where mingw packages are installed from preinstalled binaries, and `libgio-2.0-0.dll` is the shared library you are building.

When meson runs a cross-compiled executable from the build root using `wine`, WINEPATH is used to specify paths to look in for shared libraries. If `Z:\usr\i686-w64-mingw32\sys-root\mingw\bin` is added to WINEPATH first, then `libgio-2.0-0.dll` will be searched in and used from that path instead of the one just compiled from `Z:\home\\xxx\src\glib-mingw32-build`.

In the specific case this was the case and since the pre-installed older version was missing required functions, the described error occurred. Therefore, paths referring to the build directory would have to be entered first.

Reply:
> I can't even reproduce the bug you are talking about. I exported WINEPATH="Z:\\usr\\x86_64-w64-mingw32\\bin;Z:\\usr\\x86_64-w64-mingw32\\sys-root\\mingw\\bin" and built glib on Fedora against the system-wide cross-mingw64 toolchain and everything worked as expected.

This is probably due to the fact that there was no older dll with the same name in the paths mentioned. If there is, you can check this by running the command line used by Meson with `strace -e trace=file ...` and limiting the output to the specified paths.

Reply:
That doesn't seem to have anything to do with what @nirbheek said?

> If `Z:\usr\i686-w64-mingw32\sys-root\mingw\bin` is added to WINEPATH first

@nirbheek's point is that the existing code leaves this directory alone from your previous WINEPATH, it places all of self.test.extra_paths first. Which I assume is supposed to only be the build directory paths...

Anyway, your PR destroys the existing ordering of the user exported WINEPATH from `env.get('WINEPATH')` and that seems wrong.

Reply:
> it places all of self.test.extra_paths first. Which I assume is supposed to only be the build directory paths...

This is not the case. Its also adds paths provided by the compiler at https://github.com/mesonbuild/meson/blob/eafb8a8e10c1b520bb2af5b6028ea816236309d5/mesonbuild/backend/backends.py#L1048
```
                paths.update(cc.get_program_dirs(self.environment))
                paths.update(cc.get_library_dirs(self.environment))
```
This function is called according to the following call stack:
```
  File "/usr/lib/python3.6/site-packages/mesonbuild/backend/backends.py", line 380, in as_meson_exe_cmdline
    extra_paths = self.determine_windows_extra_paths(exe, extra_bdeps or [])
  File "/usr/lib/python3.6/site-packages/mesonbuild/backend/backends.py", line 817, in determine_windows_extra_paths
    result.update(self.get_mingw_extra_paths(target))
  File "/usr/lib/python3.6/site-packages/mesonbuild/backend/backends.py", line 789, in get_mingw_extra_paths
    paths.update(cc.get_program_dirs(self.environment))

Reply:
>  your PR destroys the existing ordering of the user exported WINEPATH from env.get('WINEPATH') and that seems wrong.

I took the real world example from **https://gitlab.gnome.org/GNOME/glib/-/issues/2573#note_1348456** and placed it into a simple python based test case using the sorting algorithm as in the path as shown below:
```
$ cat  test.py
build_dir = 'Z:\\home\\xxx\\src\\glib-mingw32-build'

wine_paths = [
 'Z:\\usr\\lib64\\gcc\\i686-w64-mingw32',
 'Z:\\home\\xxx\\src\\glib-mingw32-build\\gobject',
 'Z:\\usr\\i686-w64-mingw32\\sys-root\\mingw\\lib',
 'Z:\\usr\\i686-w64-mingw32\\lib',
 'Z:\\usr\\i686-w64-mingw32\\lib',
 'Z:\\usr\\i686-w64-mingw32\\lib',
 'Z:\\usr\\i686-w64-mingw32\\sys-root\\mingw\\lib',
 'Z:\\usr\\lib64\\gcc\\i686-w64-mingw32\\9.2.0',
 'Z:\\home\\xxx\\src\\glib-mingw32-build\\subprojects\\libffi\\src',
 'Z:\\usr\\i686-w64-mingw32\\bin',
 'Z:\\usr\\i686-w64-mingw32\\sys-root\\mingw\\bin',
 'Z:\\home\\xxx\\src\\glib-mingw32-build\\gio',
 'Z:\\home\\xxx\\src\\glib-mingw32-build\\glib',
 'Z:\\usr\\i686-w64-mingw32\\bin',
 'Z:\\home\\xxx\\src\\glib-mingw32-build\\gmodule',
]

wine_paths.sort(key=lambda x: not build_dir in x)

for i in wine_paths:
    print(i)
```
On running it returns:
```
$ python3 test.py
Z:\home\xxx\src\glib-mingw32-build\gobject
Z:\home\xxx\src\glib-mingw32-build\subprojects\libffi\src
Z:\home\xxx\src\glib-mingw32-build\gio
Z:\home\xxx\src\glib-mingw32-build\glib
Z:\home\xxx\src\glib-mingw32-build\gmodule
Z:\usr\lib64\gcc\i686-w64-mingw32
Z:\usr\i686-w64-mingw32\sys-root\mingw\lib
Z:\usr\i686-w64-mingw32\lib
Z:\usr\i686-w64-mingw32\lib
Z:\usr\i686-w64-mingw32\lib
Z:\usr\i686-w64-mingw32\sys-root\mingw\lib
Z:\usr\lib64\gcc\i686-w64-mingw32\9.2.0
Z:\usr\i686-w64-mingw32\bin
Z:\usr\i686-w64-mingw32\sys-root\mingw\bin
Z:\usr\i686-w64-mingw32\bin
```
You can see, that the used sorting algorithm divides the different paths into two groups, in build dir and outside. Within a group, the path order is not changed.

Reply:
> Which I assume is supposed to only be the build directory paths...

I retried building glib2 for mingw64 on opensuse with  
```
$ rpm -q meson
meson-0.54.2-lp152.2.3.1.noarch
```
after I added dumping the paths from `exe.extra_paths` to the unmodified meson installation and got 
```
/usr/x86_64-w64-mingw32/sys-root/mingw/lib
/usr/lib64/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/lib/../lib
/usr/x86_64-w64-mingw32/bin
/usr/lib64/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin
/usr/lib64/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/lib
/home/xxx/src/glib-mingw64-build/gobject
/home/xxx/src/glib-mingw64-build/subprojects/libffi/src
/usr/lib64/gcc/x86_64-w64-mingw32/9.2.0
/home/xxx/src/glib-mingw64-build/gmodule
/home/xxx/src/glib-mingw64-build/gio
/usr/lib64/gcc/x86_64-w64-mingw32
/usr/x86_64-w64-mingw32/sys-root/mingw/lib/../lib
/home/xxx/src/glib-mingw64-build/glib
/usr/x86_64-w64-mingw32/lib
```

As you can see paths pointing to the system are at not at the end of the list.